crates/kitsune_p2p/fetch/src/pool.rs

```diff
@@ -280,20 +266,24 @@ impl<'a> Iterator for StateIter<'a> {
             .take(NUM_ITEMS_PER_POLL)
             .cloned()
             .collect();
+
         for key in keys {
             let item = self.state.queue.get_refresh(&key)?;
             let item_not_recently_fetched = item
                 .last_fetch
                 .map(|t| t.elapsed() >= self.config.item_retry_delay())
-                .unwrap_or(true);
+                .unwrap_or(true); // true on the first fetch before `last_fetch` is set
             if item_not_recently_fetched {
+                // Sources are tested separately, no need to retest here
                 if let Some(source) = item.sources.next(self.config.source_retry_delay()) {
+                    // TODO what if we're recently tried to use this source and it's not available? The retry delay does not apply across items
```

Comment:
This is relevant along with the failing test I've pushed up on this PR. We are only backing off sources within each fetch item. We are not backing off sources across all fetch items. To put that in real terms, if a host is offline and we have a queue of things to fetch from that host then we will try them for all the items we are trying to fetch. The trouble with fixing that is that we use async messaging to get the data we want back so we don't know if that is going to fail. Is this dealt with elsewhere rather than here? It doesn't look like the fetch task deals with it so I'm not sure.

Reply:
https://github.com/holochain/holochain/pull/2643#discussion_r1287561566

Reply:
I've addressed the other comment and although they're related I still think this is relevant. It doesn't look like we have a mechanism to skip sources that aren't responding across the pool, only per item.

Reply:
Ah, yes, I was stuck on the multiple sources in one item, and forgot to consider across items. Indeed that is something we may want to consider addressing, we'd just want to be careful not to add too much complexity.

Reply:
I remember grappling with this problem when first writing this and deciding to get something simpler out sooner and leave room for improvement later. The problem is it's hard to remember where all the places are where there is room for improvement after a bunch of time has passed. So I'm glad you're revisiting this, this is a good opportunity to see if this would be a useful improvement given the extra complexity.

Reply:
I think this is fairly minor improvement for a reasonable amount of complexity. We'd have to make the fetch task send a limited number of notifies, wait for some responses to come back, then send some more, if the host is still responding send some more etc. But because we're using a messaging pattern, I doubt it'll give us a big performance improvement because we won't be blocking for that long on each send and the responses come back separately.

What this might be good for is reducing a bit of bandwidth usage if we can demonstrate that Kitsune will keep trying these sources when a host goes offline.

For now, I'm happy that this is a valid idea to explore and I'll leave the TODO comment in so that it's easier to get a reminder of this at some point in the future